### PR TITLE
Add arbitrary (inequality) constraints to ParameterSpace

### LINKED ISSUE #2343: Support ParameterSpaces with (inequality) constraints

### PR DESCRIPTION
**Pull request by sdrave**
This adds a `constraints` parameter to `ParameterSpace.__init__` and `Parameters.space`, that allows defining arbitrary inequality constraints in addition to the box constraints specified by the `ranges` parameter. The sampling methods are adapted accordingly.

Fixes #2343.

### CODE REVIEW

**Comment by sdrave** [comment]
@lrenelt, would that work for you?

**Comment by lrenelt** [comment]
Thanks for taking care of this. I see one issue with the number of the returned parameters. For the uniform sampling, there is no general way of ensuring that one receives the number of queried parameters. I think this should still be communicated, eg by logging the number of returned constrained parameters. However, for the random sampling this would be possible. I think that this should also be done, as it avoids an empty return if only few parameters are drawn with restrictive constraints.

**Comment by sdrave** [comment]
@lrenelt, for random sampling, what you have requested is implemented already in the PR, right? I would add a warning in case the sampling gets stuck for very long.

For uniform sampling, idk. Always logging the number of sampled parameters seems too verbose to me. Also, for these methods, the number provided by the user never is the number of parameters returned (except for 1d parameter spaces). I think I'd only add a warning, when the sample set is empty.

**Comment by lrenelt** [comment]
> @lrenelt, for random sampling, what you have requested is implemented already in the PR, right? I would add a warning in case the sampling gets stuck for very long.

True, my bad.

**Comment by lrenelt** [comment]
Oh, and there is a typo you might want to fix before merging ;) `unconstrainted -> unconstrained`

**Comment by sdrave** [comment]
I have added documentation but decided against logging warnings for now to keep things simple.

File: src/pymor/parameters/base.py [hunk 1]
```diff
@@ -616,12 +622,18 @@ def __init__(self, parameters, *ranges):
...
-        self.parameters = parameters
+        assert constraints is None or callable(constraints)
+        self.__auto_init(locals())
         self.ranges = SortedFrozenDict((k, tuple(v)) for k, v in ranges.items())
 
     def sample_uniformly(self, counts):
         """Uniformly sample |parameter values| from the space.
 
+        In the case of additional :attr:`~ParameterSpace.constraints`, the samples
+        are generated w.r.t. the box constraints specified by
```
Comment (HenKlei) [inline src/pymor/parameters/base.py:633]: The constraints might not only be "box" constraints, right?

**Review by HenKlei — APPROVED** [review]
Looks good to me!

File: src/pymor/parameters/base.py (hunk 1 above)
Comment (sdrave) [reply src/pymor/parameters/base.py:633]: I think either you didn't read the sentence completely, or I don't understand what you are saying ..

Comment (HenKlei) [reply src/pymor/parameters/base.py:633]: Of course I did not read the sentence correctly, sorry.

**Pull request merged**